src/Functions/CountSubstringsImpl.h

```diff
@@ -80,7 +80,10 @@ struct CountSubstringsImpl
                 continue;
             }
             pos = begin + haystack_offsets[i];
-            ++i;
+            if (i + 1 < input_rows_count)
```

Comment:
Why not like this?
```
++i;
if (...)
   break;
```

Also, there is no need for `chassert` here due to the added check.